torch_sim/models/metatensor.py

```diff
@@ -0,0 +1,328 @@
+                            ),
+                        ),
+                        components=[
+                            Labels(
+                                ["xyz"],
+                                torch.tensor([[0], [1], [2]], device=self._device),
+                            )
+                        ],
+                        properties=Labels(
+                            ["distance"], torch.tensor([[0]], device=self._device)
+                        ),
+                    ),
+                )
+
+            systems.append(system)
+
+        # Get model output
+        model_outputs = self._model(
+            systems=systems,
+            options=self._evaluation_options,
+            check_consistency=self._check_consistency,
+        )
+
+        results = {}
+        results["energy"] = model_outputs["energy"].block().values.detach().squeeze(-1)
+
+        # Compute forces and/or stresses if requested
+        tensors_for_autograd = []
+        if self._compute_forces:
+            for system in systems:
+                tensors_for_autograd.append(system.positions)  # noqa: PERF401
+        if self._compute_stress:
+            for strain in strains:
+                tensors_for_autograd.append(strain)  # noqa: PERF402
+
+        if self._compute_forces or self._compute_stress:
+            derivatives = torch.autograd.grad(
+                outputs=model_outputs["energy"].block().values,
+                inputs=tensors_for_autograd,
+                grad_outputs=torch.ones_like(model_outputs["energy"].block().values),
+            )
+        else:
+            derivatives = []
+
+        results_by_system: dict[str, list[torch.Tensor]] = {}
+        if self._compute_forces and self._compute_stress:
+            results_by_system["forces"] = [-d for d in derivatives[: len(systems)]]
+            results_by_system["stress"] = [
+                d / torch.abs(torch.det(system.cell.detach()))
+                for d, system in zip(derivatives[len(systems) :], systems, strict=False)
+            ]
+        elif self._compute_forces:
+            results_by_system["forces"] = [-d for d in derivatives]
+        elif self._compute_stress:
+            results_by_system["stress"] = [
+                d / torch.abs(torch.det(system.cell.detach()))
+                for d, system in zip(derivatives, systems, strict=False)
+            ]
+        else:
+            pass
+
+        # Concatenate/stack forces and stresses
+        if self._compute_forces:
+            if len(results_by_system["forces"]) > 0:
+                results["forces"] = torch.cat(results_by_system["forces"])
+            else:
+                results["forces"] = torch.empty_like(positions)
+        if self._compute_stress:
+            if len(results_by_system["stress"]) > 0:
+                results["stress"] = torch.cat(results_by_system["stress"])
+            else:
+                results["stress"] = torch.empty_like(cell)
```

Comment:
May be wrong because I am not sure about the shapes of the tensors in `results_by_system` but I believe that the forces should be concatenated and the stresses should be stacked here. Tests should catch it either way.

Reply:
Absolutely, this was a bug on my side, thanks for pointing it out. No, the tests weren't catching it because I'm seeing that, during the tests, the model is always called on a single structure

Reply:
The consistency tests won't catch it up the validate model output tests should. If not, I need to fix the tests 🙂

```python
test_mattersim_model_outputs = make_validate_model_outputs_test(
    model_fixture_name="metatensor_model",
)
```